stream/stream.go

```diff
@@ -4,229 +4,230 @@ package stream
-		case <-ctx.Done():
-			return ctx.Err()
-		}
+		c.m.RecvCounter.WithLabelValues("cache").Inc()
 	}
-	return ctx.Err()
 }
 
 func (s *Client) BuilderBid() <-chan structs.BuilderBidExtended {
 	return s.builderBidOut
 }
 
-func (s *Client) RunBuilderBidSubscriber(ctx context.Context) error {
-	l := s.Logger.WithField("method", "runBuilderBidSubscriber")
-	var bb structs.BuilderBidExtended
+func (c *Client) subscribe(ps PubSub, handle func(transport.Message)) {
+	s := c.Bids.Subscribe(c.ctx)
+	defer s.Close()
 
-	for raw := range s.builderBidIn {
-		receivedAt := time.Now()
-		sData, err := s.decode(raw)
+	for {
+		msg, err := s.Next(c.ctx)
 		if err != nil {
-			l.WithError(err).Warn("failed to decode builder bid  wrapper")
-			continue
+			return // subscription only returns fatal errors
 		}
 
-		if sData.Meta().Source == s.Config.ID {
-			continue
-		}
+		handle(msg)
+	}
+}
 
-		switch forkEncoding := sData.Meta().ForkEncoding; forkEncoding {
-		case BellatrixJson:
-			var bbb bellatrix.BuilderBidExtended
-			if err := json.Unmarshal(sData.Data(), &bbb); err != nil {
-				l.WithError(err).WithField("forkEncoding", forkEncoding).Warn("failed to decode builder bid")
-				continue
-			}
-			bb = &bbb
-		case CapellaJson:
-			var cbb capella.BuilderBidExtended
-			if err := json.Unmarshal(sData.Data(), &cbb); err != nil {
-				l.WithError(err).WithField("forkEncoding", forkEncoding).Warn("failed to decode builder bid")
-				continue
-			}
-			bb = &cbb
-		case CapellaSSZ:
-			var cbb capella.BuilderBidExtended
-			if err := cbb.UnmarshalSSZ(sData.Data()); err != nil {
-				l.WithError(err).WithField("forkEncoding", forkEncoding).Warn("failed to decode builder bid")
-				continue
-			}
-			bb = &cbb
-		default:
-			l.WithField("forkEncoding", forkEncoding).Warn("unkown builder bid forkEncoding")
-			continue
+func (c *Client) handleBid(msg transport.Message) {
+	var (
+		receivedAt = time.Now()
+		bb         structs.BuilderBidExtended
+	)
+
+	switch forkEncoding := msg.ForkEncoding; forkEncoding {
+	case transport.BellatrixJson:
+		var bbb bellatrix.BuilderBidExtended
+		if err := json.Unmarshal(msg.Payload, &bbb); err != nil {
+			c.Logger.WithError(err).
+				WithField("method", "runBuilderBidSubscriber").
+				WithField("forkEncoding", forkEncoding).
+				Warn("failed to decode builder bid")
+			return
+		}
+		bb = &bbb
+	case transport.CapellaJson:
+		var cbb capella.BuilderBidExtended
+		if err := json.Unmarshal(msg.Payload, &cbb); err != nil {
+			c.Logger.WithError(err).
+				WithField("method", "runBuilderBidSubscriber").
+				WithField("forkEncoding", forkEncoding).
+				Warn("failed to decode builder bid")
+			return
+		}
+		bb = &cbb
+	case transport.CapellaSSZ:
+		var cbb capella.BuilderBidExtended
+		if err := cbb.UnmarshalSSZ(msg.Payload); err != nil {
+			c.Logger.WithError(err).
+				WithField("method", "runBuilderBidSubscriber").
+				WithField("forkEncoding", forkEncoding).
+				Warn("failed to decode builder bid")
+			return
```

Comment:
```suggestion
func (c *Client) handleBid(msg transport.Message) error {
	var (
		receivedAt = time.Now()
		bb         structs.BuilderBidExtended
	)

	switch forkEncoding := msg.ForkEncoding; forkEncoding {
	case transport.BellatrixJson:
		var bbb bellatrix.BuilderBidExtended
		if err := json.Unmarshal(msg.Payload, &bbb); err != nil {
			return err
		}
		bb = &bbb
	case transport.CapellaJson:
		var cbb capella.BuilderBidExtended
		if err := json.Unmarshal(msg.Payload, &cbb); err != nil {
			return err
		}
		bb = &cbb
	case transport.CapellaSSZ:
		var cbb capella.BuilderBidExtended
		if err := cbb.UnmarshalSSZ(msg.Payload); err != nil {
			return err
```
Nothing important but by returning you can just log it once in subscribe. removing all of this